### PR TITLE
Ensure worker process shutdown calls atexit handlers

### PR DESCRIPTION
**Pull request by jcrist**
Processes managed by Python's `multiprocessing` don't follow the normal
interpreter shutdown procedure by default (since `multiprocessing` kills
workers with `SIGTERM`). We register our own signal handler to fix that,
ensuring that the normal shutdown procedure is followed (including
calling atexit hooks). This matters both for users of the
multiprocessing scheduler, as well as distributed's `LocalCluster`,
since that also uses `multiprocessing` to manage workers (and also uses
`dask.multiprocessing.initialize_worker_process`).

### CODE REVIEW

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks Jim! 😄

As we have typically been on a 2 week release cycle, was hoping we might have more time to test these changes before a release. Sorry things got accelerated

**Review by jrbourbeau — COMMENTED** [review]
Thanks @jcrist! FWIW `test_worker_shutdown_follows_normal_shutdown_process` passes on `main` for me locally. Should I expect it to only fail on certain platforms/Python versions/etc?

**Comment by jcrist** [comment]
Hmmm, you're right. This PR does fix a bug we found, but this test isn't sufficient it seems. Will investigate.

**Comment by jcrist** [comment]
Ah, looks like `ProcessPoolExecutor` does this automatically (while `multiprocessing.Pool` or `multiprocessing.Process` does not). The bug still affects `LocalCluster` though. I'm going to modify the test just to check that `initialize_worker_process` does this.

**Comment by jakirkham** [comment]
In that case would it make sense to only do this when working with `multiprocessing.Pool`?

**Comment by jcrist** [comment]
I don't think it's harmful, but since this would only affect the `LocalCluster` with the new `ProcessPoolExecutor` based local scheduler, perhaps this fix should be in distributed instead. There's no way to apply this fix to an already existing `multiprocessing.Pool` (since the initializer must be applied at init time).

**Comment by jcrist** [comment]
Yeah, moving this over to `distributed` instead makes more sense, given the `ProcessPoolExecutor` changes the local multiprocessing scheduler is no longer affected by this issue.

**Comment by jakirkham** [comment]
Ah I see. Meaning [here]( https://github.com/dask/distributed/blob/20a55e91b52d8e51a62ba1b47ccc1ece07adb72e/distributed/nanny.py#L723 )?

**Comment by jcrist** [comment]
Actually, there's nothing to do for distributed either, things already work there as well. This was a bit of a masqueraded issue coupled with some things that used to be true but no longer are:

- The old `multiprocessing` scheduler (using `multiprocessing.Pool`) did have this issue. We patched around it in prefect and moved on.
- The distributed `LocalCluster` was still displaying this issue, and was fixed by a similar patch I made locally to distributed. When writing tests though, I was unable to get the test to fail with no changes to distributed (meaning the fix didn't do what I thought it did). Turns out the bug is in prefect, and only happened to be fixed due to a slight difference in shutdown procedure ordering.

In summary, nothing to do here now that dask isn't using `multiprocessing.Pool`. The lingering bug is in prefect not distributed.

**Comment by jakirkham** [comment]
Oh that's great news! Thanks for the detailed summary and investigation 😄

Well if you do spot other issues, please let us know 🙂
